src/FeedForwardNeuralNetwork.cpp

```diff
@@ -92,72 +113,118 @@ void FeedForwardNeuralNetwork::setBeta(const int &ib, const double &beta)
+
 }
 
 
-double FeedForwardNeuralNetwork::getVariationalParameter(const int &i)
+void FeedForwardNeuralNetwork::randomizeBetas()
 {
-   // variables
-   double vp;
-   std::vector<NNLayer *>::size_type il = 1;
-   bool flag = false;
-   // find and store the variational parameter with the given index i
-   while ( (!flag) && (il<_L.size()) )
-   {
-      flag = _L[il]->getVariationalParameter(i,vp);
-      il++;
-   }
-   // check for errors
-   if (!flag)
-   {
-      using namespace std;
-      cout << endl << "ERROR FeedForwardNeuralNetwork::getVariationalParameter : variational parameter with the given id (" << i<< ") was not found." << endl << endl;
-   }
-   // return the result
-   return vp;
+    std::random_device rdev;
+    std::mt19937_64 rgen = std::mt19937_64(rdev());
+    std::uniform_real_distribution<double> rd = std::uniform_real_distribution<double>(-3.,3.);
+
+    // set betas to new random values
+    for (int i=0; i<this->getNBeta(); ++i) this->setBeta(i, rd(rgen));
 }
 
 
+
 // --- Computation
 
 double FeedForwardNeuralNetwork::getVariationalFirstDerivative(const int &i, const int &iv1d)
 {
-   return ( _L[_L.size()-1]->getUnit(i)->getVariationalFirstDerivativeValue(iv1d) );
+   return ( _L.back()->getUnit(i+1)->getVariationalFirstDerivativeValue(iv1d) );
+}
+
+
+void FeedForwardNeuralNetwork::getVariationalFirstDerivative(double ** vd1)
+{
+    for (int i=0; i<getNOutput(); ++i){
+        for (int iv1d=0; iv1d<getNBeta(); ++iv1d){
+            vd1[i][iv1d] = getVariationalFirstDerivative(i, iv1d);
+        }
+    }
 }
 
 
 double FeedForwardNeuralNetwork::getSecondDerivative(const int &i, const int &i2d)
 {
-   return ( _L[_L.size()-1]->getUnit(i)->getSecondDerivativeValue(i2d) );
+   return ( _L.back()->getUnit(i+1)->getSecondDerivativeValue(i2d) );
+}
+
+
+void FeedForwardNeuralNetwork::getSecondDerivative(double ** d2)
+{
+    for (int i=0; i<getNOutput(); ++i){
+        for (int i2d=0; i2d<getNInput(); ++i2d){
+            d2[i][i2d] = getSecondDerivative(i, i2d);
+        }
+    }
 }
 
 
 double FeedForwardNeuralNetwork::getFirstDerivative(const int &i, const int &i1d)
 {
-   return ( _L[_L.size()-1]->getUnit(i)->getFirstDerivativeValue(i1d) );
+   return ( _L.back()->getUnit(i+1)->getFirstDerivativeValue(i1d) );
+}
+
+
+void FeedForwardNeuralNetwork::getFirstDerivative(double ** d1)
+{
+    for (int i=0; i<getNOutput(); ++i){
+        for (int i1d=0; i1d<getNInput(); ++i1d){
+            d1[i][i1d] = getFirstDerivative(i, i1d);
+        }
+    }
 }
 
 
 double FeedForwardNeuralNetwork::getOutput(const int &i)
 {
-   return ( _L[_L.size()-1]->getUnit(i)->getValue() );
+   return _L.back()->getUnit(i+1)->getValue();
+}
+
+
+void FeedForwardNeuralNetwork::getOutput(double * out)
+{
+    for (int i=1; i<_L.back()->getNUnits(); ++i){
+        out[i-1] = _L.back()->getUnit(i)->getValue();
+    }
+}
+
+
+void FeedForwardNeuralNetwork::evaluate(const double * in, double * out, double ** d1, double ** d2, double ** vd1){
```

Comment:
I would  default the derivs to NULL and check against for it, so that I may call this without having deriv pointers, although the ffnn has derivs set. I think I will just do this small edit before merging.